### ISSUE TITLE
Bug: MapAttribute elements cannot have null=True

### ISSUE THREAD

**Issue by jl-DaDar** [pynamodb/PynamoDB#240]
I get the following error when I have MapAttribute where some of the element in it is null=True.
and if I don't set the element, the element is null : True on dynamodb, and the model then fails to deserialize.

I set element null=True because I want some map element to be optional. Is this wrong?

Exception Type: TypeError at /app/v1/user/address/
Exception Value: 'bool' object is not iterable
Request information:

File "/home/ec2-user/py3env/local/lib/python3.5/site-packages/pynamodb/models.py" in get
  470.                 return cls.from_raw_data(item_data)

File "/home/ec2-user/py3env/local/lib/python3.5/site-packages/pynamodb/models.py" in from_raw_data
  503.                 kwargs[attr_name] = attr.deserialize(attr.get_value(value))

File "/home/ec2-user/py3env/local/lib/python3.5/site-packages/pynamodb/attributes.py" in deserialize
  583.             deserialized_lst.append(class_for_deserialize.deserialize(attr_value))

File "/home/ec2-user/py3env/local/lib/python3.5/site-packages/pynamodb/attributes.py" in deserialize
  493.             deserialized_dict[key] = attr_class.deserialize(attr_value)

File "/home/ec2-user/py3env/local/lib/python3.5/site-packages/pynamodb/attributes.py" in deserialize
  488.         for k in values:

**Comment by cooldar** [comment]
looks like it due to empty mapattribute within another mapattribute
#217

**Comment by Lizards** [comment]
I've also run into this issue.  A nullable value on a MapAttribute set to None gets deserialized to True when the item is retrieved from DynamoDB.  Here are sample models and some output showing more detail:

Models:
```python
class NullableDemo(Model):
    """ OK: bar attribute can be None """
    class Meta:
        table_name = 'NullableDemo'
        host = 'http://172.17.0.2:8000'

    name = UnicodeAttribute(hash_key=True)
    foo = UnicodeAttribute(default='foo')
    bar = UnicodeAttribute(null=True)


class NullableDemoAttribute(MapAttribute):
    foo = UnicodeAttribute(default='foo')
    bar = UnicodeAttribute(null=True)


class NullableDemoWithMap(Model):
    """ Not OK: custom_map.bar deserializes to True when saved as None """
    class Meta:
        table_name = 'NullableDemoWithMap'
        host = 'http://172.17.0.2:8000'

    name = UnicodeAttribute(hash_key=True)
    custom_map = NullableDemoAttribute()`
```

Here's shell output showing the attribute values on the model instances as well as the raw output from DynamoDB via boto3:
```
In [1]: import boto3
   ...: from test import NullableDemo, NullableDemoWithMap, NullableDemoAttribute
   ...: dynamodb = boto3.client('dynamodb', region_name='us-east-1', endpoint_url='http://172.17.0.2:8000')
   ...: 

In [2]: nullable_ok = NullableDemo('demo')
   ...: nullable_ok.save()
   ...: nullable_ok = NullableDemo.get('demo')
   ...: nullable_ok.foo, nullable_ok.bar
   ...: 
Out[2]: (u'foo', None)

In [3]: dynamodb.scan(TableName='NullableDemo')['Items']
   ...: 
Out[3]: [{u'foo': {u'S': u'foo'}, u'name': {u'S': u'demo'}}]

In [4]: nullable_ok.foo = 'updated foo'
   ...: nullable_ok.bar = None
   ...: nullable_ok.save()
   ...: nullable_ok = NullableDemo.get('demo')
   ...: nullable_ok.foo, nullable_ok.bar
   ...: 
Out[4]: (u'updated foo', None)

In [5]: dynamodb.scan(TableName='NullableDemo')['Items']
Out[5]: [{u'foo': {u'S': u'updated foo'}, u'name': {u'S': u'demo'}}]

In [6]: nullable_fail = NullableDemoWithMap('demo')
   ...: nullable_fail.custom_map = NullableDemoAttribute()
   ...: nullable_fail.save()
   ...: nullable_ok = NullableDemo.get('demo')
   ...: nullable_fail.custom_map.foo, nullable_fail.custom_map.bar
   ...: 
Out[6]: ('foo', None)

In [7]: dynamodb.scan(TableName='NullableDemoWithMap')['Items']
   ...: 
Out[7]: [{u'custom_map': {u'M': {u'foo': {u'S': u'foo'}}}, u'name': {u'S': u'demo'}}]

In [8]: nullable_fail.custom_map.foo = 'updated foo'
   ...: nullable_fail.custom_map.bar = None
   ...: nullable_fail.save()
   ...: nullable_fail = NullableDemoWithMap.get('demo')
   ...: nullable_fail.custom_map.foo, nullable_fail.custom_map.bar
   ...: 
Out[8]: (u'updated foo', True)

In [9]: dynamodb.scan(TableName='NullableDemoWithMap')['Items']
Out[9]: 
[{u'custom_map': {u'M': {u'bar': {u'NULL': True},
    u'foo': {u'S': u'updated foo'}}},
  u'name': {u'S': u'demo'}}]
```

**Comment by jpinner-lyft** [comment]
#217 has been fixed in version 3.2.0 -- a similar issue #309 is being fixed in #375
